### PR TITLE
[REVIEW] Add dictionary support to libcudf groupby functions

### PR DESCRIPTION
**Pull request by davidwendt**
Reference #5963 Add dictionary support to groupby.

- [x] argmax
- [x] argmin
- [x] collect
- [x] count
- [x] max
- [x] mean* 
- [x] median
- [x] min
- [x] nth element
- [x] nunique
- [x] quantile
- [x] std*
- [x] sum* 
- [x] var* 

* _not supported due to 10.2 compile segfault_

### CODE REVIEW

File: cpp/src/groupby/hash/groupby.cu [hunk 1]
```diff
@@ -105,12 +106,15 @@ flatten_single_pass_aggs(std::vector<aggregation_request> const& requests)
       col_ids.push_back(i);
     };
 
+    auto values_type = request.values.type().id() == type_id::DICTIONARY32
```
Comment (jrhemstad) [inline cpp/src/groupby/hash/groupby.cu:109]: Is there an `is_dictionary` trait that could be used instead? That insulates against future dictionary types.

Comment (davidwendt) [reply cpp/src/groupby/hash/groupby.cu:109]: No. I can add one though.

**Comment by jrhemstad** [comment]
I realized that this PR should wait until https://github.com/rapidsai/cudf/pull/6392 as it will likely have a lot of conflicts.

File: cpp/include/cudf/utilities/traits.hpp [hunk 2]
```diff
@@ -136,27 +136,6 @@ struct is_numeric_impl {
...
-  return std::is_unsigned<T>::value;
-}
-
-struct is_unsigned_impl {
-  template <typename T>
-  bool operator()()
-  {
-    return is_unsigned<T>();
-  }
-};
```
Comment (davidwendt) [inline cpp/include/cudf/utilities/traits.hpp:158]: These seemed oddly place within the `is_numeric` set of classes/expressions in the header file so I just moved it.

File: cpp/src/groupby/sort/group_quantiles.cu [hunk 3]
```diff
@@ -55,55 +93,38 @@ struct quantiles_functor {
...
-                                               selector, segment_size, q, interpolation);
-                                           });
-
-                         for (size_t j = 0; j < num_quantiles; j++) {
-                           size_type group_size = d_group_size.element<size_type>(i);
-                           if (group_size == 0)
-                             d_result.set_null(i * num_quantiles + j);
-                           else
-                             d_result.set_valid(i * num_quantiles + j);
-                         }
```
Comment (davidwendt) [inline cpp/src/groupby/sort/group_quantiles.cu:94]: Converted this lambda in to functor `calculate_quantile_fn` to make it cleaner and easier to reuse.

File: cpp/src/groupby/hash/groupby.cu [hunk 4]
```diff
@@ -438,7 +450,10 @@ std::pair<std::unique_ptr<table>, std::vector<aggregation_result>> groupby(
       groupby_null_templated<false>(keys, requests, &cache, include_null_keys, stream, mr);
   }
 
-  return std::make_pair(std::move(unique_keys), extract_results(requests, cache));
+  auto result = extract_results(requests, cache);
+
+  // return std::make_pair(std::move(unique_keys), extract_results(requests, cache));
+  return std::make_pair(std::move(unique_keys), std::move(result));
```
Comment (karthikeyann) [inline cpp/src/groupby/hash/groupby.cu:456]: What is the reason for this change?

Comment (davidwendt) [reply cpp/src/groupby/hash/groupby.cu:456]: That was for debugging something. I will change it back.

**Comment by davidwendt** [comment]
rerun tests

**Comment by davidwendt** [comment]
Another CI build error 139

**Comment by davidwendt** [comment]
rerun tests

**Comment by davidwendt** [comment]
rerun tests

**Comment by davidwendt** [comment]
rerun tests

**Comment by davidwendt** [comment]
I was able to recreate the segfault using CUDA 10.2 but was unable to find a workaround for this. 
I verified that the segfault does not occur on 10.1 or 11.0 and does not depend on gcc or Linux version.
I created an nvbug with the details here: https://nvbugswb.nvidia.com/NvBugs5/SWBug.aspx?bugid=3186317&cp=
I will remove the experimental/diagnosing code changes from this PR.

File: cpp/tests/groupby/group_argmax_test.cpp [hunk 5]
```diff
@@ -160,8 +161,56 @@ TEST_F(groupby_argmax_string_test, zero_valid_values)
...
+  auto expect_keys = cudf::dictionary::encode(expect_keys_w);
+
+  test_single_agg(
+    keys->view(), vals_w, expect_keys->view(), expect_vals, cudf::make_argmax_aggregation());
+  test_single_agg(
+    keys_w, vals->view(), expect_keys_w, expect_vals, cudf::make_argmax_aggregation());
+  test_single_agg(
+    keys->view(), vals->view(), expect_keys->view(), expect_vals, cudf::make_argmax_aggregation());
+
+  test_single_agg(keys->view(),
```
Comment (devavret) [inline cpp/tests/groupby/group_argmax_test.cpp:194]: Because the implementation separates the grouping of keys from the aggregation calculation, the test for grouping by key type is in `groupby_keys_test.cpp`. You wouldn't need to test for dictionary keys everywhere.

File: cpp/include/cudf/detail/aggregation/aggregation.cuh [hunk 6]
```diff
@@ -278,6 +278,60 @@ struct elementwise_aggregator {
...
+  operator()(mutable_column_device_view& target,
+             size_type& target_index,
+             column_device_view& d_dictionary,
+             size_type& source_index) const noexcept
+  {
+    auto const keys  = d_dictionary.child(1);
+    auto const value = keys.element<KeyType>(
+      static_cast<cudf::size_type>(d_dictionary.element<dictionary32>(source_index)));
+    using Target = target_type_t<KeyType, aggregation::SUM>;
+    atomicAdd(&target.element<Target>(target_index), static_cast<Target>(value));
```
Comment (karthikeyann) [inline cpp/include/cudf/detail/aggregation/aggregation.cuh:299]: Try to reuse the existing code with something like
```
auto const new_source_index = static_cast<cudf::size_type>(d_dictionary.element<dictionary32>(source_index));
update_target_element<KeyType, k, target_has_nulls, source_has_nulls>{}(target, target_index, keys, new_source_index);
```

**Review by karthikeyann — CHANGES_REQUESTED** [review]

File: cpp/include/cudf/detail/aggregation/aggregation.cuh (hunk 6 above)
Comment (davidwendt) [reply cpp/include/cudf/detail/aggregation/aggregation.cuh:299]: I could not get this to work. The only specialization is needed for `Source=dictionary32` and `k=SUM`. The rest of work fine since they either just count group values are find the indexes of the extrema. This means that these can just use the dictionary column's indices which works by default without any changes. The `SUM` agg needs the actual values from the child keys column so that specialization is the only one required.

File: cpp/tests/groupby/group_argmax_test.cpp (hunk 5 above)
Comment (davidwendt) [reply cpp/tests/groupby/group_argmax_test.cpp:194]: I don't think I understand. Each groupby/agg seemed to require some tweaks to the code in order to support dictionary types. Some cases only needed the indices while others required using the keys. Adding tests for each groupby/agg seemed  prudent to ensure groupby can support dictionary columns. The groupby code is very complex so I may have misunderstanding of how it works.

Comment (devavret) [reply cpp/tests/groupby/group_argmax_test.cpp:194]: Yes. There's two scenarios: 
1. the keys are dictionary columns
2. the values are dictionary columns
You only need to test for the latter in each of the test files that follow the naming pattern `group_<agg>_test.cpp`. For the former, it is sufficient to add dictionary keys test to `groupby_keys_test.cpp`.

**Comment by davidwendt** [comment]
I cannot get around the 10.2 ptxas compile segfault as documented here https://nvbugswb.nvidia.com/NvBugs5/SWBug.aspx?bugid=3186317
The code logic that seems to cause the error is isolated in the aggregation `sum` specialization for dictionary type. This means we cannot support aggregation-sum-groupby for dictionary columns as values. And aggregation-sum is also required in order to support mean, std, and var aggregations. Using a dictionary column as keys for groupby work fine.

The PR includes logic to throw an exception if a dictionary column is used as values with one of these aggregation types. The code is isolated around a CUDA==10.2 equivalent compile directive so technically these aggregations work if compiled with 10.1 or 11.0 but the gtests have been commented out.

File: cpp/src/groupby/sort/group_quantiles.cu [hunk 7]
```diff
@@ -14,26 +14,64 @@
...
+    size_type offset = i * num_quantiles;
+    thrust::for_each_n(thrust::seq,
+                       thrust::make_counting_iterator(0),
+                       num_quantiles,
+                       [d_result = d_result, segment_size, offset](size_type j) {
+                         if (segment_size == 0)
+                           d_result.set_null(offset + j);
+                         else
+                           d_result.set_valid(offset + j);
+                       });
```
Comment (jrhemstad) [inline cpp/src/groupby/sort/group_quantiles.cu:71]: This looks like it will be very slow as it is setting each bit individually. Can't this use `valid_if`?

Comment (jrhemstad) [reply cpp/src/groupby/sort/group_quantiles.cu:71]: Nevermind, I see we're inside a device function.

**Comment by davidwendt** [comment]
rerun tests

**Comment by mike-wendt** [comment]
rerun tests

File: cpp/tests/groupby/group_sum_test.cpp [hunk 8]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2019, NVIDIA CORPORATION.
+ * Copyright (c) 2019-2020, NVIDIA CORPORATION.
```
Comment (karthikeyann) [inline cpp/tests/groupby/group_sum_test.cpp:2]: in few files, year is updated, in few, it's not.
what are the rules for updating the year?

File: cpp/tests/groupby/group_nth_element_test.cpp [hunk 9]
```diff
@@ -344,5 +348,25 @@ TYPED_TEST(groupby_nth_element_test, basic_string)
...
+  using K = int32_t;
+  using V = std::string;
+
+  // clang-format off
+  fixed_width_column_wrapper<K> keys  {1,   2,   3,   1,   2,   2,   1,   3,   3,   2};
+  dictionary_column_wrapper<V>  vals{"AB", "1", "2", "3", "4", "5", "6", "7", "8", "9"};
+  // groupby =>  key:  { 1, 1, 1, 2, 2, 2, 2, 3, 3, 3};
+  //             vals: {AB, 3, 6, 1, 4, 5, 9, 2, 7, 8};
+  fixed_width_column_wrapper<K> expect_keys{ 1,   2,   3};
+  dictionary_column_wrapper<V> expect_vals_w{"6", "5", "8"};
```
Comment (karthikeyann) [inline cpp/tests/groupby/group_nth_element_test.cpp:362]: `clang-format off` isn't useful here. the code doesn't look beautified.

File: cpp/tests/groupby/group_mean_test.cpp [hunk 10]
```diff
@@ -137,5 +138,27 @@ TYPED_TEST(groupby_mean_test, null_keys_and_values)
 }
 // clang-format on
 
+struct groupby_dictionary_mean_test : public cudf::test::BaseFixture {
+};
+
+TEST_F(groupby_dictionary_mean_test, basic)
```
Comment (karthikeyann) [inline cpp/tests/groupby/group_mean_test.cpp:144]: This test seems like a placeholder until 10.2 is removed, or the bug is fixed.
if this test needs to be disabled, prefix DISABLED_ to the name.

File: cpp/tests/groupby/group_mean_test.cpp [hunk 11]
```diff
@@ -14,19 +14,20 @@
...
 #include <cudf_test/base_fixture.hpp>
 #include <cudf_test/column_wrapper.hpp>
 #include <cudf_test/type_list_utilities.hpp>
 #include <cudf_test/type_lists.hpp>
 
-#include <cudf/detail/aggregation/aggregation.hpp>
+#include <tests/groupby/groupby_test_util.hpp>
+
+#include <initializer_list>
+#include <iterator>
```
Comment (karthikeyann) [inline cpp/tests/groupby/group_mean_test.cpp:28]: ide auto-inserted?

File: cpp/src/groupby/sort/group_std.cu [hunk 12]
```diff
@@ -33,30 +35,56 @@ namespace cudf {
...
+
+template <typename ResultType, typename Iterator>
 struct var_transform {
-  column_device_view d_values;
-  column_device_view d_means;
-  column_device_view d_group_sizes;
+  column_device_view const d_values;
+  Iterator values_iter;
+  ResultType const* d_means;
+  size_type const* d_group_sizes;
```
Comment (karthikeyann) [inline cpp/src/groupby/sort/group_std.cu:44]: neat.

**Review by karthikeyann — CHANGES_REQUESTED** [review]

File: cpp/tests/groupby/group_mean_test.cpp (hunk 11 above)
Comment (davidwendt) [reply cpp/tests/groupby/group_mean_test.cpp:28]: No. I just rearranged these.

File: cpp/src/groupby/hash/groupby.cu [hunk 13]
```diff
@@ -135,6 +137,8 @@ class hash_compound_agg_finalizer final : public cudf::detail::aggregation_final
       stream(stream),
       mr(mr)
   {
+    result_type = cudf::is_dictionary(col.type()) ? cudf::dictionary_column_view(col).keys().type()
```
Comment (jrhemstad) [inline cpp/src/groupby/hash/groupby.cu:140]: For groupby, I wonder if it really makes sense for the output to also be dictionary encoded. Typically the result is relatively small and the values unique.  Dictionary encoding such a column would almost certainly just be pure overhead without any benefit.

Comment (jrhemstad) [reply cpp/src/groupby/hash/groupby.cu:140]: Oh, nevermind, it isn't dictionary encoding it. Carry on.

**Review by jrhemstad — APPROVED** [review]

File: cpp/tests/groupby/group_mean_test.cpp (hunk 10 above)
Comment (davidwendt) [reply cpp/tests/groupby/group_mean_test.cpp]: Cool. Thanks for this.

**Review by karthikeyann — APPROVED** [review]

File: cpp/tests/groupby/group_sum_test.cpp (hunk 8 above)
Comment (harrism) [reply cpp/tests/groupby/group_sum_test.cpp:2]: If you change a file, make sure the current year is included. Since this PR was started in 2020, most files didn't need a year update.

File: cpp/src/groupby/sort/group_std.cu [hunk 14]
```diff
@@ -33,30 +35,56 @@ namespace cudf {
...
+
     size_type group_idx  = d_group_labels[i];
-    size_type group_size = d_group_sizes.element<size_type>(group_idx);
+    size_type group_size = d_group_sizes[group_idx];
 
     // prevent divide by zero error
     if (group_size == 0 or group_size - ddof <= 0) return 0.0;
 
-    ResultType mean = d_means.element<ResultType>(group_idx);
+    ResultType mean = d_means[group_idx];
```
Comment (devavret) [inline cpp/src/groupby/sort/group_std.cu:60]: This is sort of the reverse direction of what I've been doing. `column_device_view`'s `element()` accessor is more helpful than simple subscript. e.g. in fixed point. `element()` will return the value with the scale applied. And that scale is stored once in the column.

Any particular reason for this change?

**Review by devavret — COMMENTED** [review]
Damn. I missed it by a minute.
